### PR TITLE
knn classifier from scratch.

### PR DESCRIPTION
**Pull request by justin212k**
DO NOT MERGE YET.

WIP knn classifier using distance matrix. Test runs, but results may be wrong.

To run:
  pip install -e .
  pytest q2_sample_classifier/tests/test_classifier.py::EstimatorsTests::test_classify_samples_from_dist -s

### CODE REVIEW

**Comment by nbokulich** [comment]
@justin212k re: your latest commit... instead of relying on `predict_and_plot` it would be better to check out the new pipelines `classify_samples` and `regress_samples` and use the actions used in those pipelines for plotting results.

**Comment by justin212k** [comment]
Only reason the test passes is because the correct result is hardcoded in classify_samples_from_dist. I think next is to 
1. make the classifier actually work for the 1-nearest-neighbor case
2. also return accuracy_results (and eventually, more)

@alenajk , take a stab if you have time?

**Comment by nbokulich** [comment]
@justin212k that's fine — if a trained estimator does not make sense, there's no need to output it (and there are some other actions, e.g., `classify_samples_ncv` that do not output the trained estimator). Right now these pipelines are mostly used stand-alone, though there are a couple other pipelines in QIIME2 (e.g., sample-classifier's `maturity_index`) that wrap these pipelines and use the trained estimator... so just a few cases where we would not be able to add this pipeline as an option.

It looks like this is becoming a separate pipeline, so it would make sense to keep the inputs/outputs as consistent as possible with `classify_samples`, but really feel free to add/drop outputs as makes sense for this pipeline.

**Comment by justin212k** [comment]
sg, thanks

**Comment by nbokulich** [comment]
@justin212k I will review! Thanks for getting this in!

Before I review, can you make sure tests pass? If coverage is decreasing for a good reason, pls just let me know. Thanks!

**Comment by justin212k** [comment]
hey, it's green! I'm not sure how to interpret the coverage results though.

**Comment by justin212k** [comment]
mmm, merge conflicts. Should I try to resolve?

**Comment by nbokulich** [comment]
@justin212k I just resolved — let's see if that works.

sorry, a different PR was merged this morning that introduced those conflicts.

**Comment by justin212k** [comment]
roger

File: q2_sample_classifier/classify.py [hunk 1]
```diff
@@ -38,6 +41,47 @@
 }
 
 
+def classify_samples_from_dist(ctx, dmtx, metadata, k,
+                               palette=defaults['palette']):
+    ''' Returns knn classifier results from a distance matrix.'''
+    distance_matrix = dmtx.view(skbio.DistanceMatrix)
+    predictions = []
+    metadata_series = metadata.to_series()
+    for i, row in enumerate(distance_matrix):
```
Comment (nbokulich) [inline q2_sample_classifier/classify.py]: it looks like you can just use [condensed_form](http://scikit-bio.org/docs/0.5.1/generated/generated/skbio.stats.distance.DistanceMatrix.condensed_form.html#skbio.stats.distance.DistanceMatrix.condensed_form) to replicate this operation (then use `reindex` and `dropna` to resort and subset the metadata categories). Right?

File: q2_sample_classifier/classify.py [hunk 2]
```diff
@@ -38,6 +41,47 @@
 }
 
 
+def classify_samples_from_dist(ctx, dmtx, metadata, k,
```
[the remark this answers is not in the document]
Comment (nbokulich) [reply q2_sample_classifier/classify.py]: also should k have a default value?

File: q2_sample_classifier/classify.py [hunk 3]
```diff
@@ -38,6 +41,47 @@
...
+    for i, row in enumerate(distance_matrix):
+        dists = []
+        categories = []
+        for j, dist in enumerate(row):
+            if j == i:
+                continue  # exclude self
+            dists.append(dist)
+            categories.append(metadata_series[distance_matrix.ids[j]])
+
+        # k-long series of (category: dist) ordered small -> large
```
Comment (nbokulich) [inline q2_sample_classifier/classify.py]: oh cool — looks like you just built your own knn classifier from scratch. Any reason not to use sklearn?

File: q2_sample_classifier/classify.py [hunk 4]
```diff
@@ -334,6 +378,11 @@ def summarize(output_dir: str, sample_estimator: Pipeline):
     _summarize_estimator(output_dir, sample_estimator)
 
 
+def summarize_knn(output_dir: str, k: int):
+    params = pd.Series({'k': 1}, name='Parameter setting')
```
Comment (nbokulich) [inline q2_sample_classifier/classify.py]: k appears to be hardcoded here so the summary always displays k=1

File: q2_sample_classifier/plugin_setup.py [hunk 5]
```diff
@@ -226,6 +228,37 @@
...
+        'palette': 'The color palette to use for plotting.',
+        },
+    output_descriptions={
+        'predictions': 'leave one out predictions for each sample',
+        'model_summary': 'number of neighbors k',
+        'accuracy_results': 'Accuracy results visualization.',
+    },
+    name=('Train and test a cross-validated (leave one out)'
+          ' supervised learning classifier.'),
+    description=description.format(
```
Comment (nbokulich) [inline q2_sample_classifier/plugin_setup.py]: this pipeline should probably have its own description instead of recycling the standard description — in particular, it would be useful to mention that this uses LOO and briefly describe the significance of K.

**Comment by nbokulich** [comment]
oh also: looks like an unused function was reimported when I fixed the merge conflicts — please remove when you make your changes. Thanks!

File: q2_sample_classifier/classify.py (hunk 3 above)
Comment (justin212k) [reply q2_sample_classifier/classify.py]: yeah... Don't love the reinvention of knn here. But seems that sklearn's knn doesn't accept a distance matrix [1], and I want this to be usable with e.g. unifrac.

Could reach into the more private parts of sklearn though, or better would probably be submit a PR to sklearn with knn_from_distance_matrix

[1] http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html#sklearn.neighbors.KNeighborsClassifier

**Comment by justin212k** [comment]
re: general comments:
model summary removed. Auto-tuning k and a knn regressor sound good to me.

Any of your tests use unifrac or weighted unifrac? I was hoping the badness of knn relative to RF would be offset by the phylogenetic information from unifrac.

File: q2_sample_classifier/classify.py (hunk 1 above)
Comment (justin212k) [reply q2_sample_classifier/classify.py]: Don't immediately see how, though I admit this is a pretty dumb brute force approach. condensed_form on the distance_matrix would require that we keep track of which pairs each value belongs to, yes?

Might be a little while until I can look at this more, so feel free to edit if you like.

File: q2_sample_classifier/classify.py (hunk 3 above)
Comment (nbokulich) [reply q2_sample_classifier/classify.py]: if metric == precomputed, it takes a distance matrix according to that link

But I'm not sure exactly how this would work for prediction.

It's a fairly simple method, so I don't think there's a problem with the reinvention here... this can be a baseline that we can use to implement sklearn's knn classifier later on if we see some advantage.

**Comment by nbokulich** [comment]
My tests were all with unifrac... but it was just to give it a spin. I will save a more comprehensive benchmark for later.

The low quality probably has more to do with the specific datasets and distance metrics (e.g., weighted UniFrac might have performed better with those data)

File: q2_sample_classifier/classify.py (hunk 1 above)
Comment (nbokulich) [reply q2_sample_classifier/classify.py]: I think you're right about needing to keep track of pairs... but I have not given it a try. Seems like we should be able to get that info from the raw indices.

Nothing wrong with brute force... just wondering if there's a more efficient approach.

File: q2_sample_classifier/plugin_setup.py [hunk 6]
```diff
@@ -249,11 +248,11 @@
         },
     output_descriptions={
         'predictions': 'leave one out predictions for each sample',
-        'model_summary': 'number of neighbors k',
         'accuracy_results': 'Accuracy results visualization.',
     },
-    name=('Train and test a cross-validated (leave one out)'
-          ' supervised learning classifier.'),
+    name=('Run k-nearest-neighbors on a labeled distance matrix.'
```
Comment (nbokulich) [inline q2_sample_classifier/plugin_setup.py]: oh the "name" looked fine, though this looks better!

My comment about the description referred to the "description" below — you can more or less just replicate the "name" in the "description". Currently, the description line below prints this in the help docs:
> Predicts a categorical sample metadata column using a supervised learning
  classifier. Splits input data into training and test sets. The training
  set is used to train and test the estimator using a stratified k-fold
  cross-validation scheme. This includes optional steps for automated
  feature extraction and hyperparameter optimization. The test set validates
  classification accuracy of the optimized estimator. Outputs classification
  results for test set. For more details on the learning algorithm, see
  http://scikit-learn.org/stable/supervised_learning.html

File: q2_sample_classifier/classify.py (hunk 3 above)
Comment (justin212k) [reply q2_sample_classifier/classify.py]: Aye, it may be just my unfamiliarity with using sklearn's knn.

**Comment by justin212k** [comment]
sorry, didn't realize I left this in a broken state. Hopefully tests will pass now

**Comment by justin212k** [comment]
Yay!
